pkg/apiserver/user/sqlauth/sqlauth.go

```diff
@@ -50,13 +67,123 @@ func (a *Authenticator) Authenticate(f user.AuthenticateForm) (*utils.SessionUse
 	}
 	defer utils.CloseTiDBConnection(db) //nolint:errcheck
 
+	// Check dashboard privileges
+	// 1. Check whether TiDB SEM is enabled
+	resData, err := a.tidbClient.SendGetRequest("/config")
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	var config tidbSecurityConfig
+	err = json.Unmarshal(resData, &config)
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	// 2. Get grants
+	var grantRows []string
+	err = db.Raw("show grants for current_user()").Find(&grantRows).Error
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	grants := parseCurUserGrants(grantRows)
+	// 3. Check
+	if !checkDashboardPriv(grants, config.Security.EnableSem) {
+		return nil, user.ErrLackPrivileges.NewWithNoMessage()
+	}
+
 	return &utils.SessionUser{
 		Version:      utils.SessionVersion,
 		HasTiDBAuth:  true,
 		TiDBUsername: f.Username,
 		TiDBPassword: f.Password,
 		DisplayName:  f.Username,
 		IsShareable:  true,
-		IsWriteable:  true,
+		IsWriteable:  checkWriteablePriv(grants),
 	}, nil
 }
+
+// Currently, There are 2 kinds of grant output format in TiDB:
+// - GRANT [grants] ON [db.table] TO [user]
+// - GRANT [roles] TO [user]
+// Examples:
+// - GRANT PROCESS,SHOW DATABASES,CONFIG ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT SYSTEM_VARIABLES_ADMIN,RESTRICTED_VARIABLES_ADMIN,RESTRICTED_STATUS_ADMIN,RESTRICTED_TABLES_ADMIN ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT ALL PRIVILEGES ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT `app_read`@`%` TO `test`@`%`
+func parseCurUserGrants(grantRows []string) set {
+	grants := set{}
+	grantRegex := regexp.MustCompile(`GRANT (.+) ON`)
+
+	for _, row := range grantRows {
+		m := grantRegex.FindStringSubmatch(row)
+		if len(m) == 2 {
+			curRowGrants := strings.Split(m[1], ",")
+			for _, grant := range curRowGrants {
+				grants[grant] = struct{}{}
+			}
+		}
+	}
+
+	return grants
+}
+
+// To access TiDB Dashboard, following base privileges are required
+// - ALL PRIVILEGES
+// - or
+// - PROCESS
+// - SHOW DATABASES
+// - CONFIG
+// - DASHBOARD_CLIENT or SUPER (SUPER includes DASHBOARD_CLIENT)
+// When TiDB SEM is enabled, following extra privileges are required
+// - RESTRICTED_VARIABLES_ADMIN
+// - RESTRICTED_TABLES_ADMIN
+// - RESTRICTED_STATUS_ADMIN
+func checkDashboardPriv(privs set, enableSEM bool) bool {
+	hasPriv := func(priv string) bool {
+		_, ok := privs[priv]
+		return ok
+	}
+
+	if !hasPriv("ALL PRIVILEGES") {
+		if !hasPriv("PROCESS") {
+			return false
+		}
+		if !hasPriv("SHOW DATABASES") {
+			return false
+		}
+		if !hasPriv("CONFIG") {
+			return false
+		}
+		if !hasPriv("SUPER") {
+			if !hasPriv("DASHBOARD_CLIENT") {
+				return false
+			}
+		}
+	}
+	if enableSEM {
+		if !hasPriv("RESTRICTED_VARIABLES_ADMIN") {
+			return false
+		}
+		if !hasPriv("RESTRICTED_TABLES_ADMIN") {
+			return false
+		}
+		if !hasPriv("RESTRICTED_TABLES_ADMIN") {
+			return false
+		}
```

Comment:
How about reorg this part to the first? Then the overall indention can be reduced:

```go
if enableSEM {
	// Note: When SEM is enabled, these additional privileges need to be checked even if "ALL PRIVILEGES" is granted.
	if !hasPriv("RESTRICTED_VARIABLES_ADMIN") {
		return false
	}
	if !hasPriv("RESTRICTED_TABLES_ADMIN") {
		return false
	}
	if !hasPriv("RESTRICTED_TABLES_ADMIN") {
		return false
	}
}

if hasPriv("ALL PRIVILEGES") {
	// ALL PRIVILEGES contains privileges below. If it is set, privilege requirement is met.
	return true
}

if !hasPriv("PROCESS") {
	return false
}
if !hasPriv("SHOW DATABASES") {
	return false
}
if !hasPriv("CONFIG") {
	return false
}

if hasPriv("SUPER") {
	// SUPER contains privileges below. If it is set, privilege requirement is met.
	return true
}

if !hasPriv("DASHBOARD_CLIENT") {
	return false
}
```

Reply:
Done.